### PR TITLE
feat: multi-objective REA

### PR DESCRIPTION
**Pull request by aaronkl**
Adds a multi-objective version of regularized evolution.

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: syne_tune/optimizer/schedulers/searchers/multi_objective_regularized_evolution.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    RegularizedEvolution,
+)
+
+from syne_tune.optimizer.schedulers.multiobjective.multiobjective_priority import (
+    MOPriority,
+    NonDominatedPriority,
+)
+
+
+class MultiObjectiveRegularizedEvolution(RegularizedEvolution):
```
Comment (geoalgo) [inline syne_tune/optimizer/schedulers/searchers/multi_objective_regularized_evolution.py:29]: It seems to me that we could also just allow to use a different sort function in regularized evolution, I mean replace score by multiobjective priorities and compute priorities whenever a decision is made. Let me know what your thoughts on this, the advantage would be less code but if you think it is too convoluted or difficult to do we can also do this approach.

File: tst/schedulers/test_schedulers_api.py [hunk 2]
```diff
@@ -361,11 +369,14 @@ def make_transfer_learning_evaluations(num_evals: int = 10):
...
     trial_ids = range(4)
 
-    if isinstance(scheduler, MOASHA):
+    if isinstance(scheduler, MOASHA) or isinstance(scheduler, MOREA):
         assert scheduler.metric_names() == [metric1, metric2]
     else:
         assert scheduler.metric_names() == [metric1]
-    assert scheduler.metric_mode() == mode
+
+    print(scheduler, mode)
```
Comment (geoalgo) [inline tst/schedulers/test_schedulers_api.py:377]: Can you remove this line?
```suggestion
```

File: syne_tune/optimizer/schedulers/searchers/multi_objective_regularized_evolution.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    Adapts regularized evolution algorithm by Real et al. to the multi-objective setting. Elements in the
+    populations are scored via non-dominated sorting. Parents are sampled from the population based on
```
Comment (geoalgo) [inline syne_tune/optimizer/schedulers/searchers/multi_objective_regularized_evolution.py:32]:
```suggestion
    populations are scored via a multi-objective priority that is set to non-dominated sort by default. Parents are sampled from the population based on
```

**Review by geoalgo — COMMENTED** [review]
LGTM, I just have a question whether we can provide an option to just use a custom score in REA.

File: syne_tune/optimizer/schedulers/searchers/multi_objective_regularized_evolution.py (hunk 1 above)
Comment (geoalgo) [reply syne_tune/optimizer/schedulers/searchers/multi_objective_regularized_evolution.py:28]: To be clear, this is non blocking and I am fine merging as is. In my opinion, having a custom sort would be slightly better: there would be less code and I think it is quite easy to explain to a user that the algorithm can be customized with different criterion, some of which are multi-objective (you could also have a subclass which just instantiate a specific non-dominated sort as the priority function).

We can also do it in two steps, i.e. merge now and do the custom sort, later as you prefer.

Comment (aaronkl) [reply syne_tune/optimizer/schedulers/searchers/multi_objective_regularized_evolution.py:28]: Ok then let's do as you suggested

**Review by geoalgo — APPROVED** [review]

**Pull request merged**